test/parallel_api/algorithm/alg.sorting/sort_by_key_common.h

```diff
@@ -0,0 +1,293 @@
+}
+
+template<typename Policy, typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_sort(Policy&& policy, KeyIt keys_begin, ValIt vals_begin, Size n, Compare compare, UnstableSortTag)
+{
+    if constexpr (std::is_same_v<Compare, NoComparator>)
+        oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+    else
+        oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin, compare);
+}
+
+template<typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_reference_sort(KeyIt ref_keys_begin, ValIt ref_vals_begin, Size n, Compare compare)
+{
+    auto first = oneapi::dpl::make_zip_iterator(ref_keys_begin, ref_vals_begin);
+    if constexpr (std::is_same_v<Compare, NoComparator>)
+    {
+        std::stable_sort(first, first + n, [](const auto& lhs, const auto& rhs) {
+            return std::get<0>(lhs) < std::get<0>(rhs);
+        });
+    }
+    else
+    {
+        std::stable_sort(first, first + n, [compare](const auto& lhs, const auto& rhs) {
+            return compare(std::get<0>(lhs), std::get<0>(rhs));
+        });
+    }
+}
+
+template<typename KeyIt, typename ValIt, typename KeysOrigIt, typename ValsOrigIt, typename Size, typename Compare>
+void
+check_sort(const KeyIt& keys_begin, const ValIt& vals_begin,
+           const KeysOrigIt& keys_orig_begin, const ValsOrigIt& vals_orig_begin,
+           Size keys_n, Size vals_n, Compare compare, StableSortTag)
+{
+    using KeyT = typename std::iterator_traits<KeyIt>::value_type;
+    using ValT = typename std::iterator_traits<ValIt>::value_type;
+    std::vector<KeyT> keys_expected(keys_orig_begin, keys_orig_begin + keys_n);
+    std::vector<ValT> vals_expected(vals_orig_begin, vals_orig_begin + vals_n);
+    call_reference_sort(keys_expected.begin(), vals_expected.begin(), keys_n, compare);
+    EXPECT_EQ_N(keys_expected.begin(), keys_begin, keys_n, "wrong result stable sort: keys");
+    EXPECT_EQ_N(vals_expected.begin(), vals_begin, keys_n, "wrong result stable sort: values");
+    EXPECT_EQ_N(vals_expected.begin() + keys_n, vals_begin + keys_n, vals_n - keys_n,
+                "wrong result stable sort: remaining values should not be touched");
+}
+
+template<typename KeyIt, typename ValIt, typename KeysOrigIt, typename ValsOrigIt, typename Size, typename Compare>
+void
+check_sort(const KeyIt& keys_begin, const ValIt& vals_begin,
+           const KeysOrigIt& keys_orig_begin, const ValsOrigIt& vals_orig_begin,
+           Size keys_n, Size vals_n, Compare compare, UnstableSortTag)
+{
+    using KeyT = typename std::iterator_traits<KeyIt>::value_type;
+    using ValT = typename std::iterator_traits<ValIt>::value_type;
+    std::vector<KeyT> keys_expected(keys_orig_begin, keys_orig_begin + keys_n);
+    std::vector<ValT> vals_expected(vals_orig_begin, vals_orig_begin + vals_n);
+    call_reference_sort(keys_expected.begin(), vals_expected.begin(), keys_n, compare);
```

Comment:
It looks the same reference sorting (etalon) is called for stable and unstable cases...

Reply:
> Lets add //TODO about that.

Added